opm/simulators/linalg/gpuistl/set_device.cpp

```diff
@@ -17,33 +17,62 @@
   along with OPM.  If not, see <http://www.gnu.org/licenses/>.
 */
 #include <config.h>
+#include <sstream>
 #include <cuda_runtime.h>
 #include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/simulators/flow/FlowGenericVanguard.hpp>
 #include <opm/simulators/linalg/gpuistl/detail/gpu_safe_call.hpp>
 #include <opm/simulators/linalg/gpuistl/set_device.hpp>
+
 namespace Opm::gpuistl
 {
 void
 setDevice(int mpiRank, [[maybe_unused]] int numberOfMpiRanks)
 {
+    try {
+        int deviceCount = -1;
+        [[maybe_unused]] auto cuError = cudaGetDeviceCount(&deviceCount);
 
-    int deviceCount = -1;
-    [[maybe_unused]] auto cuError = cudaGetDeviceCount(&deviceCount);
+        if (deviceCount <= 0) {
+            // If they have CUDA/HIP enabled (ie. using a component that needs CUDA, eg. gpubicgstab or CUILU0), this will fail
+            // later down the line. At this point in the simulator, we can not determine if CUDA is enabled, so we can only
+            // issue a warning.
+            OpmLog::warning("Could not find any CUDA/HIP devices.");
+            return;
+        }
 
-    if (deviceCount <= 0) {
-        // If they have CUDA enabled (ie. using a component that needs CUDA, eg. gpubicgstab or CUILU0), this will fail
-        // later down the line. At this point in the simulator, we can not determine if CUDA is enabled, so we can only
-        // issue a warning.
-        OpmLog::warning("Could not find any CUDA devices.");
-        return;
+        // Now do a round robin kind of assignment
+        // TODO: We need to be more sophistacted here. We have no guarantee this will pick the correct device.
+        const auto deviceId = mpiRank % deviceCount;
+        OPM_GPU_SAFE_CALL(cudaDeviceReset());
+        OPM_GPU_SAFE_CALL(cudaSetDevice(deviceId));
     }
+    catch(...){}
```

Comment:
I don't think it's a good idea to have a catch all here. At the very least just catch `std::runtime_error` which is what `OPM_GPU_SAFE_CALL` will call, but it is probably a better idea to just call `OPM_GPU_WARN_IF_ERROR` and remove the whole try/catch.

Reply:
done